lib/src/rules/prefer_int_value_is_even_check.dart

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2020, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+import '../util/boolean_expression_utilities.dart';
+
+const _desc =
+    r'Prefer intValue.isOdd/isEven instead of checking the result of % 2.';
+
+const _details = r'''
+
+**PREFER** the use of intValue.isOdd/isEven to check for evenness.
+
+**BAD:**
+```
+bool isEven = 1 % 2==0;
+bool isOdd = 13 % 2 ==1;
+```
+
+**GOOD:**
+```
+bool isEven =1.isEven;
+bool isOdd = 13.isOdd;
+```
+
+''';
+
+class PreferIntValueIsEvenCheck extends LintRule implements NodeLintRule {
+  PreferIntValueIsEvenCheck()
+      : super(
+            name: 'prefer_int_value_is_even_check',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  void registerNodeProcessors(NodeLintRegistry registry,
+      [LinterContext context]) {
+    final visitor = _Visitor(this);
+    registry.addSimpleIdentifier(this, visitor);
+    registry.addBinaryExpression(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final LintRule rule;
+  _Visitor(this.rule);
+
+  @override
+  void visitBinaryExpression(BinaryExpression node) {
+    var left = node.leftOperand;
+    var leftType = left.staticType;
+    var right = node.rightOperand;
+    var rightType = right.staticType;
+    // Both sides have to have static type of int
+    if (!(leftType?.isDartCoreInt == true &&
+        rightType?.isDartCoreInt == true &&
+        right is IntegerLiteral)) {
+      return;
+    }
+    if (left is BinaryExpression) {
+      var rightChild = left.rightOperand;
+      var rightChildType = left.staticType;
+      if (rightChildType?.isDartCoreInt == true &&
+          rightChild is IntegerLiteral &&
+          rightChild.value == 2 &&
+          BooleanExpressionUtilities.ALL_COMPARISONS_OPERATIONS
+              .contains(node.operator.type)) {
```

Comment:
We should check the `node`'s operator at the beginning of the method because it will be more efficient than waiting until this point.

Also, we should check the `left` operand's operator to ensure that it's `%`. As written I believe that the lint rule will produce a diagnostic for `a + 2 == 0` (which should be added as a test case either way).

Reply:
Good catch. Done